compiler/rustc_lint/src/builtin.rs

```diff
@@ -2469,18 +2469,7 @@ impl<'tcx> LateLintPass<'tcx> for InvalidValue {
 
         /// Determines whether the given type is inhabited. `None` means that we don't know.
         fn ty_inhabited<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> Option<bool> {
-            use rustc_type_ir::sty::TyKind::*;
-            if !cx.tcx.type_uninhabited_from(cx.param_env.and(ty)).is_empty() {
-                // This is definitely uninhabited from some module.
-                return Some(false);
-            }
-            match ty.kind() {
-                Never => Some(false),
-                Int(_) | Uint(_) | Float(_) | Bool | Char | RawPtr(_) => Some(true),
-                // Fallback for more complicated types. (Note that `&!` might be considered
-                // uninhabited so references are "complicated", too.)
-                _ => None,
-            }
+            ty.uninhabited_predicate(cx.tcx).apply_any_module(cx.tcx, cx.param_env).map(|b| !b)
```

Comment:
What do you think of changing it to `InhabitedPredicate` to reduce double-negatives like this?

Reply:
I like it. Will that cause a lot of churn? If so, maybe in a follow up PR?

Reply:
Not much churn unless I also change `tcx.is_ty_uninhabited_from`. But I think I'd like to keep that one as it since it is usually not negated.

Reply:
sgtm then